### PR TITLE
Better types compliance

### PR DESCRIPTION
**Pull request by Hydrocharged**
First off, I'm going to list the types that are missing from this PR:

```
DECIMAL
ENUM
SET
GEOMETRY
GEOMETRYCOLLECTION
LINESTRING
MULTILINESTRING
POINT
MULTIPOINT
POLYGON
MULTIPOLYGON
```

The first three mentioned are the last ones I'm going to implement for now, as the `GEOMETRY` types are very, very complex. However, `DECIMAL` proved to be a bit tougher than expected, and I didn't want to delay getting this PR out since I'll be gone over the entire Thanksgiving holiday (technically I'm on it now but I promised a PR by the 25th). Besides the ones listed above, every other type is in this PR in some capacity (`NCHAR` and friends are missing, but that type can be replicated as it's just a charset/collation shortcut).

Some notes, because they're important. There are no tests. Changing the types has broken a fair amount, and I haven't bothered to fix those things yet before I can get started writing tests.

I'm making use of higher-level interfaces that are specific to some type or group of types. I played around with the idea of adding a new function to the `Type` interface that returned an attribute map, but decided against it since I wanted to be able to add more than raw attributes, and a map of functions just seemed too complex for the tradeoff.

Character sets and collations are complete, in that every single supported one in MySQL is here...which is why there are so many. I had many ideas on how to model this, but decided on using string constants as the base so that it's easy to reference them from outside code, and then have multiple maps for each specific attribute of a charset/collation. For example, if an implementer wants to specifically support the `hebrew` character set, then they can just reference `CharacterSet_hebrew`, rather than defining their own variable with `custom_var, err := ParseCharacterSet("hebrew")`.

JSON is actually a very complex type, so I didn't really change anything from the logic that was already present before. This does mean that we don't truly support `JSON`, but implementing it right may take 1-2 full weeks, which I don't feel is worth it right now.

It pains me to say, but in the effort of full compliancy, I am modeling a `BOOLEAN` as an `int8`. But that's how MySQL does it, so it's probably for the best.

You'll notice that some types have global variables while others don't, and that's because some types don't make sense to reference without other information. `Int8` can be a global type because it is fully self-contained in its information. `Varchar`, OTOH, requires a length to be meaningful.

On the topic of the string types, if any logic looks weird, then it's to mimic something that MySQL is doing. I did a lot of testing to see what MySQL actually did when the documentation wasn't clear, so some rules seem inconsistent by MySQL is apparently inconsistent. I'm also doing nothing with the character set and collation stuff regarding comparisons, so we just support `utf8mb4` as a character set (which mimics Go's string type implementation to the best of my knowledge), and changing it to something else doesn't change any actual behavior. That can come later because that's A LOT to do...

The `TIME` type is weird, in that the minimum precision is microseconds, so I chose to base the entire type around that. All of the parseable variations are present as well. I was initially going to use `time.Duration` from Go, but it actually didn't suit the requirements. Working with it turned out to be more difficult than just using microseconds. For example, the only way to make a duration is to parse a string of a specific format, which would be parsing a string, then recreating a different string to parse, which is too much.

`YEAR` is the most useless type I've ever encountered in anything, but hey it's included and it's fully supported now, hooray...

### CODE REVIEW

File: sql/textbintype.go [hunk 1]
```diff
@@ -0,0 +1,290 @@
+package sql
```
Comment (zachmu) [inline sql/textbintype.go]: Probably should name this file string_type.go

File: sql/numbertype.go [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+	Uint64 = MustCreateNumberType(sqltypes.Uint64)
+	// Float32 is a floating point number of 32 bits.
+	Float32 = MustCreateNumberType(sqltypes.Float32)
+	// Float64 is a floating point number of 64 bits.
+	Float64 = MustCreateNumberType(sqltypes.Float64)
+)
+
+type NumberType interface {
+	Type
+	IsUnsigned() bool
```
[the remark this answers is not in the document]
Comment (Hydrocharged) [reply sql/numbertype.go]: I technically don't need `IsSigned`, but I added it just for clarity sake. `if t.IsSigned` looks nicer than `if !t.IsUnsigned`, but I can remove it.

**Comment by Hydrocharged** [comment]
@zachmu My 2nd commit has all of the changes since your first review.

File: sql/bit.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/spf13/cast"
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/proto/query"
+)
+
+type BitType interface {
```
Comment (zachmu) [inline sql/bit.go]: For all these types, you should have a type comment that explains what the type represents and maybe a link to the MySQL docs for it.

File: sql/bit_test.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+)
+
+func TestBitCompare(t *testing.T) {
+	tests := []struct {
+		typ Type
+		val1 interface{}
+		val2 interface{}
+		expectedCmp int
+	}{
+		{MustCreateBitType(1), 0, 1, -1},
```
Comment (zachmu) [inline sql/bit_test.go]: No tests of nil?

File: sql/datetimetype.go [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+
+	// Date is a date with day, month and year.
+	Date = MustCreateDatetimeType(sqltypes.Date)
+	// Datetime is a date and a time
+	Datetime = MustCreateDatetimeType(sqltypes.Datetime)
+	// Timestamp is an UNIX timestamp.
+	Timestamp = MustCreateDatetimeType(sqltypes.Timestamp)
+)
+
+type DatetimeType interface {
```
Comment (zachmu) [inline sql/datetimetype.go]: Type comment

File: sql/datetimetype.go [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+	return t.baseType
+}
+
+func (t datetimeType) Zero() interface{} {
+	return zeroTime
+}
+
+// ValidateTime receives a time and returns either that time or nil if it's
+// not a valid time.
+func ValidateTime(t time.Time) interface{} {
```
Comment (zachmu) [inline sql/datetimetype.go]: Why isn't the return type time.Time?

File: sql/datetimetype_test.go [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+)
+
+func TestDatetimeCompare(t *testing.T) {
+	tests := []struct {
+		typ Type
+		val1 interface{}
+		val2 interface{}
+		expectedCmp int
+	}{
+		{Date, time.Date(2012, 12, 12, 12, 12, 12, 12, time.UTC),
```
Comment (zachmu) [inline sql/datetimetype_test.go]: Nils?

File: sql/datetimetype_test.go [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+func TestDatetimeCreate(t *testing.T) {
+	tests := []struct {
+		baseType query.Type
+		expectedType datetimeType
+		expectedErr bool
+	}{
+		{sqltypes.Date, datetimeType{sqltypes.Date}, false},
+		{sqltypes.Datetime, datetimeType{sqltypes.Datetime}, false},
+		{sqltypes.Timestamp, datetimeType{sqltypes.Timestamp}, false},
+		{sqltypes.Binary, datetimeType{}, true},
```
Comment (zachmu) [inline sql/datetimetype_test.go]: I would separate this out into two tests, one for compatible and one for incompatible types.

File: sql/expression/boolean.go [hunk 9]
```diff
@@ -28,21 +26,41 @@ func (e *Not) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
 	if err != nil {
 		return nil, err
 	}
-
 	if v == nil {
 		return nil, nil
 	}
 
+	//// if it's in an expected format then we just negate it
```
Comment (zachmu) [inline sql/expression/boolean.go]: Remove commented code?

File: sql/expression/boolean.go [hunk 10]
```diff
@@ -28,21 +26,41 @@ func (e *Not) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
...
+	//} else if b, ok := v.(bool); ok {
+	//	return !b, nil
+	//} else {
+	//	return nil, sql.ErrInvalidType.New(reflect.TypeOf(v).String())
+	//}
+
 	b, ok := v.(bool)
 	if !ok {
-		v, _ = e.Type().Convert(v)
+		v, err = sql.BooleanParse(v)
```
Comment (zachmu) [inline sql/expression/boolean.go]: Shouldn't this logic be BooleanType.Convert(v)?

File: sql/expression/convert.go [hunk 11]
```diff
@@ -24,14 +24,20 @@ const (
 	ConvertToNChar = "nchar"
 	// ConvertToDate is a conversion to date.
 	ConvertToDate = "date"
-	// ConvertToDatetime is a conversion to datetune.
+	// ConvertToDatetime is a conversion to datetine.
```
Comment (zachmu) [inline sql/expression/convert.go]: Typo

File: sql/expression/convert.go [hunk 12]
```diff
@@ -120,46 +131,51 @@ func (c *Convert) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
...
-		s, err := sql.Text.Convert(val)
-		if err != nil {
-			return nil, err
-		}
-
-		b, err := sql.Blob.Convert(s)
+		b, err := sql.Blob.Convert(val)
 		if err != nil {
-			return nil, err
+			return nil, nil
```
Comment (zachmu) [inline sql/expression/convert.go]: Err?

File: sql/expression/convert.go [hunk 13]
```diff
@@ -120,46 +131,51 @@ func (c *Convert) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
...
 		}
-
 		return b, nil
 	case ConvertToChar, ConvertToNChar:
 		s, err := sql.Text.Convert(val)
 		if err != nil {
-			return nil, err
+			return nil, nil
```
Comment (zachmu) [inline sql/expression/convert.go]: Err?

File: sql/expression/convert.go [hunk 14]
```diff
@@ -120,46 +131,51 @@ func (c *Convert) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
 func convertValue(val interface{}, castTo string) (interface{}, error) {
```
Comment (zachmu) [inline sql/expression/convert.go]: What's with the change in error handling in this function? You changed many error handling blocks to return nil on encountering an error. This is either a mistake or needs to be carefully documented.

File: sql/numbertype_test.go [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+)
+
+func TestNumberCompare(t *testing.T) {
+	tests := []struct {
+		typ Type
+		val1 interface{}
+		val2 interface{}
+		expectedCmp int
+	}{
+		{Boolean, 0, 1, -1},
```
Comment (zachmu) [inline sql/numbertype_test.go]: Nils?

File: sql/plan/show_create_table_test.go [hunk 16]
```diff
@@ -40,7 +41,7 @@ func TestShowCreateTable(t *testing.T) {
 	expected := sql.NewRow(
 		table.Name(),
 		"CREATE TABLE `test-table` (\n  `baz` text NOT NULL,\n"+
-			"  `zab` integer DEFAULT 0,\n"+
+			"  `zab` int DEFAULT 0,\n"+
```
Comment (zachmu) [inline sql/plan/show_create_table_test.go]: You should add a TODO to change the casing behavior here. It's dumb that we lowercase these names when reserved words are upper case in every other context

File: sql/stringtype_test.go [hunk 17]
```diff
@@ -0,0 +1,316 @@
...
+)
+
+func TestStringCompare(t *testing.T) {
+	tests := []struct {
+		typ Type
+		val1 interface{}
+		val2 interface{}
+		expectedCmp int
+	}{
+		{MustCreateStringWithDefaults(sqltypes.VarChar, 10), 0, 1, -1},
```
Comment (zachmu) [inline sql/stringtype_test.go]: Nils?

File: sql/stringtype_test.go [hunk 18]
```diff
@@ -0,0 +1,316 @@
...
+				assert.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				assert.Equal(t, test.expectedVal, val)
+			}
+		})
+	}
+}
+
+func TestStringString(t *testing.T) {
```
Comment (zachmu) [inline sql/stringtype_test.go]: You should go to sing sing for this stringstring

File: sql/timetype.go [hunk 19]
```diff
@@ -0,0 +1,284 @@
...
+	"gopkg.in/src-d/go-errors.v1"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/proto/query"
+)
+
+var (
```
Comment (zachmu) [inline sql/timetype.go]: It's a little strange / inconsistent that this top-level type file doesn't have a new interface type in it. Why is that? Seems like there should be a Time interface, even if just embeds the other datetime interface.

File: sql/timetype_test.go [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+		}
+		return d
+	}
+
+	tests := []struct {
+		val1 interface{}
+		val2 interface{}
+		expectedCmp int
+	}{
+		{-1, 1, -1},
```
Comment (zachmu) [inline sql/timetype_test.go]: Nils?

File: sql/type_test.go [hunk 21]
```diff
@@ -245,15 +242,13 @@ func TestNumberComparison(t *testing.T) {
...
-	require := require.New(t)
-
-	var f = numberT{
-		t: query.Type_FLOAT64,
-	}
-	val, err := f.SQL(23.222)
-	require.NoError(err)
-	require.True(val.IsFloat())
-	require.Equal(sqltypes.NewFloat64(23.222), val)
+	//TODO: rewrite this
```
Comment (zachmu) [inline sql/type_test.go]: Remove commented code?

File: sql/type_test.go [hunk 22]
```diff
@@ -12,9 +12,8 @@ import (
 func TestIsNull(t *testing.T) {
```
Comment (zachmu) [inline sql/type_test.go]: Seems like this file should be broken up to parallel the breakup of type.go, right?

File: sql/yeartype.go [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+	"vitess.io/vitess/go/vt/proto/query"
+)
+
+var (
+	Year yearType
+
+	ErrConvertingToYear = errors.NewKind("value %v is not a valid Year")
+)
+
+type yearType struct{}
```
Comment (zachmu) [inline sql/yeartype.go]: Need a top-level type comment somewhere. Also same comment as before: why isn't YearType an interface?

File: sql/plan/show_create_table_test.go (hunk 16 above)
Comment (Hydrocharged) [reply sql/plan/show_create_table_test.go]: I just fixed it instead, they were calling `ToLower` so I just removed it.

**Pull request merged**